Draft EIP: Multiple contenthash records for ENS

filips123 opened the pull request:
This adds my draft EIP for multiple `contenthash` records for ENS. Contract implementation is available in ensdomains/resolvers#30, and discussion should be in ethereum/EIPs#2393.

filips123 commented:
I'm pinging @decanus and @Arachnid as authors of original EIP-1577.

filips123 commented:
I think EIP should be mostly complete, but one thing that needs to be addressed is how to use fixed size bytes for the key (https://github.com/ethereum/EIPs/issues/2393#issuecomment-588514520). There might also be some required changes for implementation, but this is probably separate from specification PR.

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+---
+eip: <to be assigned>
```
```suggestion
eip: 2520
```
Be sure to rename the file to `EIPS/eip-2520.md` as well.

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+---
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+This EIP introduces support for multiple `contenthash` records for ENS. It allows hosting a website on multiple distributed systems and accessing it from the same ENS domain. It is an extension of the EIP 1577.
```
Simple summary should be thought of like a subject line, GitHub commit message or PR title, forum thread sub-heading, etc.
```suggestion
ENS support for multiple `contenthash` records on a single ENS name.
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+This EIP introduces support for multiple `contenthash` records for ENS. It allows hosting a website on multiple distributed systems and accessing it from the same ENS domain. It is an extension of the EIP 1577.
+
+## Motivation
+<!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+However, the domain can store only one `contenthash` record which means that the site owner needs to decide which hosting system to use. Because there are many ENS-compatible hosting systems available (IPFS, Swarm, recently Onion and ZeroNet), and there will probably be even more in the future, lack of support for multiple records could become problematic. Instead, domains should be able to store ...
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+* If the application supports multiple hosting systems (like MetaMask), it SHOULD request `contenthash` without a specific type (like in EIP 1577). The contract MUST then return the default `contenthash` record.
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised  ...
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+The proposed implementation was chosen because it is simple to implement and supports all important requested features. However, it doesn't support multiple records for the same type and priority order, as they aren't so important and are harder to implement properly. To implement them, a different way of implementation would need to be used.
+
+## Backwards Compatibility
+<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+The EIP is backwards-compatible with EIP 1577, the only difference is additional overloaded methods. Old applications will still be able to function correctly, as they will receive the default `contenthash` record.
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+Automatic Truffle tests have also been written along with new contract code. However, additional manual tests are also needed.
+
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+Implementation of the smart contract, as well as automatic tests, are written in [`ensdomains/resolvers#30`](https://github.com/ensdomains/resolvers/pull/30).
+
+## Security Considerations
+<!--All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, a ...
+
```
```suggestion
```

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+## Security Considerations
+<!--All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, a ...
+
+There SHOULD NOT be any security problems with this EIP. However, it SHOULD still be audited to check for any possible issues with using `mapping` and `bytes` for storage, and overloading methods, as well as other implementation-related features.
```
EIPs should be written targeting a future reader (5 to 10 years) long after the EIP is final.  If this EIP is not audited before then, then it probably never will be audited so commenting that someone should audit this EIP should be removed.  It may be valuable warn implementers about security related gotchas if they exist in this section.

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+Implementation of the smart contract, as well as automatic tests, are written in [`ensdomains/resolvers#30`](https://github.com/ensdomains/resolvers/pull/30).
```
Can this code be inlined into this EIP or included as an asset in `../assets/eip-2520/`?  If not, then it may be better to just remove the Implementation section (it isn't required) entirely.  We try to avoid external links in EIPs whenever possible, and in this case it isn't critical to include this in the EIP.

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+
+Automatic Truffle tests have also been written along with new contract code. However, additional manual tests are also needed.
```
Tests should either be included here (inline or as an asset in `../assets/eip-2520/`) or this section should just be left out entirely.

An ideal test case for an EIP just defines input and expected output pairs, but does not the test infrastructure or specific test tooling.

MicahZoltu commented on EIPS/eip-draft_multiple_contenthash_records.md:
```diff
@@ -0,0 +1,79 @@
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+A resolver SHOULD store `contenthash` records in `mapping`, containing `protoCode` as key and whole `contenthash` (including `protoCode`) as value. Default `contenthash` SHOULD be stored as empty `bytes` key. Alternatively, it MAY use other storage types, but public interface MUST be the same as described in this EIP.
```
Consider using `**SHOULD**` rather than `SHOULD` for consistency across this repository.  Same for all of the other RFC-2119 keywords in the document.

MicahZoltu replied:
EIPs should not be prescriptive about implementation details, so this opening statement is just adding confusion.  If you want to push for a particular implementation, that should be handled in blog posts, documentation on other sites, etc.  Ideally, the implementation sticks to the public interface only.

MicahZoltu reviewed (changes requested):
The EIP number needs to be set and the inline comments need to be removed, but the rest of the feedback can wait until a later date before being addressed if you prefer.

One thing that isn't clear to me is why this needs to be an EIP.  This feels like something that could just be documented on the ENS docs site and you can avoid the entire EIP process.

filips123 commented:
I addressed most of required changes. I deleted Test Cases and Security Considerations sections, but I will add them back if I find some important security gotchas in the future.

The reason why I decided to make this EIP is that some existing similar ENS-related specifications were also added as EIPs.

MicahZoltu commented on EIPS/eip-2520.md:
````diff
@@ -0,0 +1,72 @@
+}
+```
+
+## Copyright
````
```suggestion
## Security Considerations
TBD

## Copyright
```
Security considerations section is required.  You can leave it blank for now, and when it comes time to move to Last Call/Final if no issues are brought up you can fill it in with "There are no known security issues." or something.

MicahZoltu reviewed (changes requested):
> The reason why I decided to make this EIP is that some existing similar ENS-related specifications were also added as EIPs.

I have expressed my disagreement with most of those other people as well.  😉  We have been working on getting the EIPs repository under control lately, so have become a bit more strict about what is and is not included.  All EIP Editors are volunteer, and we found ourselves unable to keep up with the volume of EIPs so getting people to stop using the EIPs repository for stuff that doesn't need to be an EIP was one way to help alleviate that pressure.
